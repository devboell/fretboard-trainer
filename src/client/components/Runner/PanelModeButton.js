import styled from 'styled-components'

const PanelModeButton = styled.button`
  margin: 1px;
  height: 50px;
  width: 70px;

  text-align: center;
  vertical-align: middle;

  cursor: pointer;

  border-radius: 5px;
  border: 1px solid gray;
  outline: 0;
`

export default PanelModeButton
