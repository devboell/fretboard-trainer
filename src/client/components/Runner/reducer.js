import {
  compose,
  set,
  lensProp,
  lensPath,
  over,
  append,
} from 'ramda'

const INIT_RUNNER = 'INIT_RUNNER'
const RESET_RUNNER = 'RESET_RUNNER'
const RUNNER_START = 'RUNNER_START'
const PANEL_MODE_SELECTION = 'PANEL_MODE_SELECTION'
const CORRECT_ANSWER = 'CORRECT_ANSWER'
const INCORRECT_ANSWER = 'INCORRECT_ANSWER'

export const initRunner = quiz => ({
  type: INIT_RUNNER,
  quiz,
})

export const resetRunner = () => ({
  type: RESET_RUNNER,
})

export const startRunner = question => ({
  type: RUNNER_START,
  question,
})

export const selectPanelModeIdx = idx => ({
  type: PANEL_MODE_SELECTION,
  idx,
})

export const addCorrectAnswer = answer => ({
  type: CORRECT_ANSWER,
  answer,
})

export const addIncorrectAnswer = answer => ({
  type: INCORRECT_ANSWER,
  answer,
})

const initialAnswers = {
  correct: [],
  incorrect: [],
}

export const initialState = {
  quiz: undefined,
  question: undefined,
  selectedPanelModeIdx: 0,
  answers: initialAnswers,
}

export default (state = initialState, action) => {
  switch (action.type) {
    case INIT_RUNNER:
      return compose(
        set(lensProp('quiz'), action.quiz),
        set(lensProp('question'), undefined),
        set(lensProp('selectedPanelModeIdx'), 0),
        set(lensProp('answers'), initialAnswers),
      )(state)

    case RESET_RUNNER:
      return set(lensProp('quiz'), undefined, state)

    case RUNNER_START:
      return compose(
        set(lensProp('answers'), initialAnswers),
        set(lensProp('question'), action.question),
      )(state)

    case PANEL_MODE_SELECTION:
      return set(lensProp('selectedPanelModeIdx'), action.idx, state)

    case CORRECT_ANSWER: {
      const correctLens = lensPath(['answers', 'correct'])
      return over(correctLens, append(action.answer), state)
    }

    case INCORRECT_ANSWER: {
      const incorrectLens = lensPath(['answers', 'incorrect'])
      return over(incorrectLens, append(action.answer), state)
    }

    default:
      return state
  }
}
