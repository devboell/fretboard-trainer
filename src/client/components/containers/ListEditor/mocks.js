import React from 'react'
import TestProvider from 'test-utils'
import { createStore, combineReducers, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

import QUIZZES from 'graphql/Quizzes'
import PANEL_MODES from 'graphql/PanelModes'
import CREATE_QUIZ from 'graphql/CreateQuiz'
import UPDATE_QUIZ from 'graphql/UpdateQuiz'
import DELETE_QUIZ from 'graphql/DeleteQuiz'

import * as quizFxt from 'fixtures/graphql/quiz'
import * as panelFxt from 'fixtures/graphql/panelMode'

import runnerReducer, { initialState as initialRunnerState } from 'components/containers/Runner/reducer'
import listReducer, { initialState as initialListState } from 'components/containers/QuizList/reducer'
import editorReducer, { initialState as initialEditorState } from './EditorContainer/reducer'

import ListEditor from './index'


export const quizzes = {
  request: { query: QUIZZES },
  result: { data: { quizzes: quizFxt.quizzes } },
}

export const panelModes = {
  request: { query: PANEL_MODES },
  result: { data: { panelModes: panelFxt.panelModes } },
}

export const createQuiz = {
  request: {
    query: CREATE_QUIZ,
    variables: quizFxt.createQuizInputValues,
  },
  result: { data: { createQuiz: quizFxt.createdQuiz } },
}

export const updateQuiz = {
  request: {
    query: UPDATE_QUIZ,
    variables: quizFxt.updateQuizInputValues,
  },
  result: { data: { updateQuiz: quizFxt.updatedQuiz } },
}

export const deleteQuiz = {
  request: {
    query: DELETE_QUIZ,
    variables: { id: '3' },
  },
  result: { data: { deleteQuiz: '3' } },
}

export const dataMocks = [
  quizzes,
  panelModes,
  createQuiz,
  updateQuiz,
  deleteQuiz,
]

const initialState = {
  list: initialListState,
  editor: initialEditorState,
  runner: initialRunnerState,
}

export const store = createStore(
  combineReducers({
    list: listReducer,
    editor: editorReducer,
    runner: runnerReducer,
  }),
  initialState,
  applyMiddleware(thunk),
)

export const getWrapper = async () => {
  const wrapper = mount((
    <TestProvider
      store={store}
      mocks={dataMocks}
    >
      <ListEditor />
    </TestProvider>
  ))
  // makes sure the loading is done
  await new Promise(resolve => setTimeout(resolve))
  wrapper.update()

  return wrapper
}
