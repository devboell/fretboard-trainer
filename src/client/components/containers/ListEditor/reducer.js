import { reducer as listReducer } from 'components/reusable/ListDetail'
import editorReducer from 'components/containers/QuizEditor/reducer'

export const initialState = {}

export default (state = initialState, action) => ({
  list: listReducer('listEditor')(state.list, action),
  editor: editorReducer('listEditor')(state.editor, action),
})
