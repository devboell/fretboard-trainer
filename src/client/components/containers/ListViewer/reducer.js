import listReducer from 'components/containers/QuizList/reducer'

export const initialState = {}

export default (state = initialState, action) => ({
  list: listReducer('listViewer')(state.list, action),
})
