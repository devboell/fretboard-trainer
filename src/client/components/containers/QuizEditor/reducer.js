import {
  set,
  over,
  lensProp,
  lensPath,
  compose,
  append,
  without,
  pluck,
  assoc,
  dissoc,
} from 'ramda'

import { ITEM_SELECTION, CLEAR_SELECTION } from 'components/reusable/ListDetail'

const NEW_ITEM_SELECTION = 'NEW_ITEM_SELECTION'
const ITEM_UPDATE = 'ITEM_UPDATE'
const BUFFER_UPDATE = 'BUFFER_UPDATE'


export const selectNewItem = quizType => ({
  type: NEW_ITEM_SELECTION,
  quizType,
})


export const updateItem = quiz => ({
  type: ITEM_UPDATE,
  quiz,
})

export const updateBuffer = (key, value, updateType) => ({
  type: BUFFER_UPDATE,
  key,
  value,
  updateType,
})


export const modes = {
  UNSELECTED: 'UNSELECTED',
  SELECTED: 'SELECTED',
  NEW: 'NEW',
}

export const updateTypes = {
  SET: 'SET',
  ADD: 'ADD',
  REMOVE: 'REMOVE',
}

const newQuiz = type => ({
  id: undefined,
  name: '',
  description: '',
  type,
  tuning: 'standard',
  width: 13,
  panelModeIds: ['1'],
  allAnswers: false,
  allowIncorrect: false,
  useTimer: true,
  time: 5,
  showNotes: false,
  __typename: 'Quiz',
})

export const initialState = {
  original: undefined,
  buffer: undefined,
  mode: modes.UNSELECTED,
}

const convertPanelModes = (quiz) => {
  const panelModeIds = pluck('id', quiz.panelModes)
  return compose(
    assoc('panelModeIds', panelModeIds),
    dissoc('panelModes'),
  )(quiz)
}

const initBuffer = quiz => state => compose(
  set(lensProp('buffer'), quiz),
  set(lensProp('original'), quiz),
)(state)


const handleSelection = (quiz, mode) => state => compose(
  initBuffer(quiz),
  set(lensProp('mode'), mode),
)(state)

export default name => (state = initialState, action) => {
  switch (action.type) {
    /*
    case INIT_EDITOR: {
      const editableQuiz = convertPanelModes(action.quiz)
      return handleSelection(editableQuiz, modes.SELECTED)(state)
    }
    */
    case ITEM_SELECTION + name: {
      const editableQuiz = convertPanelModes(action.item)
      return handleSelection(editableQuiz, modes.SELECTED)(state)
    }

    case NEW_ITEM_SELECTION:
      return handleSelection(newQuiz(action.quizType), modes.NEW)(state)

    case CLEAR_SELECTION + name:
      return handleSelection(undefined, modes.UNSELECTED)(state)

    case ITEM_UPDATE: {
      const editableQuiz = convertPanelModes(action.quiz)
      return initBuffer(editableQuiz)(state)
    }

    case BUFFER_UPDATE: {
      const { key, value, updateType } = action
      const lensKey = lensPath(['buffer', key])

      switch (updateType) {
        case updateTypes.SET: {
          return set(lensKey, value, state)
        }
        case updateTypes.ADD: {
          return over(lensKey, append(value), state)
        }

        case updateTypes.REMOVE: {
          return over(lensKey, without([value]), state)
        }

        default:
          return state
      }
    }

    default:
      return state
  }
}
