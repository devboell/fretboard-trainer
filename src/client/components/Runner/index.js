import React from 'react'
import pt from 'prop-types'
import Fretboard from 'react-fretboard'
import { connect } from 'react-redux'
import quizShape from 'propShapes/quiz'
import getQuestion from 'lib/question'

import PanelMode from 'components/PanelMode'
import PanelModeButton from './PanelModeButton'
import StartButton from './StartButton'
import { startRunner } from './reducer'

const Runner = ({ quiz, question, onStartRunner }) =>
  <div>
    <p>{quiz.name}</p>
    <div>
      {quiz.panelModes.map(pm =>
        <PanelModeButton key={`panelModeButton=${pm.id}`}>
          <PanelMode panelMode={pm} />
        </PanelModeButton>)
      }
    </div>
    {question &&
      <div>
        <p>{question.entity.name}</p>
        <Fretboard
          skinType="strings"
          showNotes
          selectedLocations={[question.entity.rootLoc]}
        />
      </div>
    }
    <StartButton onClick={onStartRunner}>
      Start
    </StartButton>
  </div>

Runner.propTypes = {
  quiz: quizShape.isRequired,
  question: pt.shape({}),
  onStartRunner: pt.func.isRequired,
}

Runner.defaultProps = {
  question: undefined,
}

const mapStateToProps = state => ({
  quiz: state.runner.quiz,
  question: state.runner.question,
})

const mapDispatchToProps = dispatch => ({
  onStartRunner: question => dispatch(startRunner(question)),
})

const mergeProps = (stateProps, dispatchProps) => ({
  ...stateProps,
  ...dispatchProps,
  onStartRunner: () => dispatchProps.onStartRunner(getQuestion(stateProps.quiz)),
})

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Runner)
