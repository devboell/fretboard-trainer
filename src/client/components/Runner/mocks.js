import { createStore, combineReducers } from 'redux'
import reducer, { initialState } from './reducer'


export const store = createStore(combineReducers({ runner: reducer }), initialState)
export const bla = 'bla'
