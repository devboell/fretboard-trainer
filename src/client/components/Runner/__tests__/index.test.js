import React from 'react'
import { Provider } from 'react-redux'
// import * as getQuestion from 'lib/question'

import { quizzes } from 'fixtures/graphql/quiz'
import { clickStart } from 'test-utils/enzyme-queries/runner'
import Runner from '../index'
import { store } from '../mocks'
import { initRunner } from '../reducer'


store.dispatch(initRunner(quizzes[0]))

jest.mock('lib/question', () => {
  const q = {
    choices: [
      'G#',
      'E',
      'D',
      'G',
    ],
    entity: {
      name: 'D',
      otherLocs: [],
      rootLoc: {
        pos: 5,
        str: 1,
      },
    },
  }
  return jest.fn(() => q)
})

it('should behave...', () => {
  const wrapper = mount((
    <Provider store={store}>
      <Runner />
    </Provider>))

  // jest.spyOn(question, 'getQuestion')
  // question.getQuestion.mockImplementation(() => q)
  clickStart(wrapper)
  expect(wrapper).toMatchSnapshot()
})
