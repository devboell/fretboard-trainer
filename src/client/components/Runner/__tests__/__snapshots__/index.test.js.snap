// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should behave... 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <Connect(Runner)>
    <Runner
      onStartRunner={[Function]}
      question={
        Object {
          "choices": Array [
            "G#",
            "E",
            "D",
            "G",
          ],
          "entity": Object {
            "name": "D",
            "otherLocs": Array [],
            "rootLoc": Object {
              "pos": 5,
              "str": 1,
            },
          },
        }
      }
      quiz={
        Object {
          "__typename": "Quiz",
          "id": "1",
          "name": "quiz one",
          "panelModes": Array [
            Object {
              "answer": "name",
              "id": "1",
              "question": "fretboard",
            },
            Object {
              "answer": "staff",
              "id": "2",
              "question": "fretboard",
            },
            Object {
              "answer": "sound",
              "id": "3",
              "question": "fretboard",
            },
          ],
          "tuning": "standard",
          "type": "pc",
          "width": 13,
        }
      }
    >
      <div>
        <p>
          quiz one
        </p>
        <div>
          <PanelModeButton
            key="panelModeButton=1"
          >
            <button
              className="PanelModeButton-ceptcP kdnJCL"
            >
              <PanelMode
                panelMode={
                  Object {
                    "answer": "name",
                    "id": "1",
                    "question": "fretboard",
                  }
                }
              >
                <Wrapper>
                  <div
                    className="Wrapper-ThVjl daBGdj"
                  >
                    <p>
                      fretboard
                    </p>
                    <hr />
                    <p>
                      name
                    </p>
                  </div>
                </Wrapper>
              </PanelMode>
            </button>
          </PanelModeButton>
          <PanelModeButton
            key="panelModeButton=2"
          >
            <button
              className="PanelModeButton-ceptcP kdnJCL"
            >
              <PanelMode
                panelMode={
                  Object {
                    "answer": "staff",
                    "id": "2",
                    "question": "fretboard",
                  }
                }
              >
                <Wrapper>
                  <div
                    className="Wrapper-ThVjl daBGdj"
                  >
                    <p>
                      fretboard
                    </p>
                    <hr />
                    <p>
                      staff
                    </p>
                  </div>
                </Wrapper>
              </PanelMode>
            </button>
          </PanelModeButton>
          <PanelModeButton
            key="panelModeButton=3"
          >
            <button
              className="PanelModeButton-ceptcP kdnJCL"
            >
              <PanelMode
                panelMode={
                  Object {
                    "answer": "sound",
                    "id": "3",
                    "question": "fretboard",
                  }
                }
              >
                <Wrapper>
                  <div
                    className="Wrapper-ThVjl daBGdj"
                  >
                    <p>
                      fretboard
                    </p>
                    <hr />
                    <p>
                      sound
                    </p>
                  </div>
                </Wrapper>
              </PanelMode>
            </button>
          </PanelModeButton>
        </div>
        <p>
          D
        </p>
        <StartButton
          onClick={[Function]}
        >
          <button
            className="StartButton-bXgdxw iuifDU"
            onClick={[Function]}
          >
            Start
          </button>
        </StartButton>
      </div>
    </Runner>
  </Connect(Runner)>
</Provider>
`;
